lightning/restore/restore.go

```diff
@@ -307,48 +308,209 @@ outside:
+			worker.wg.Wait()
 		}
+		// for quit
+		worker.throw(nil)
+	}
+}
 
-		// restore views. Since views can cross database we must restore views after all table schemas are restored.
-		for _, dbMeta := range rc.dbMetas {
-			if len(dbMeta.Views) > 0 {
-				task := log.With(zap.String("db", dbMeta.Name)).Begin(zap.InfoLevel, "restore view schema")
-				viewsSchema := make(map[string]string)
-				for _, viewMeta := range dbMeta.Views {
-					viewsSchema[viewMeta.Name] = viewMeta.GetSchema(ctx, rc.store)
+func (worker *restoreSchemaWorker) doJob() {
+	for {
+		select {
+		case <-worker.ctx.Done():
+			return
+		case job := <-worker.jobCh:
+			if job == nil {
+				return
+			}
+			var logger log.Logger
+			for _, stmt := range job.stmts {
+				if stmt.stmtType == schemaCreateDatabase {
+					logger = log.With(zap.String("db", job.dbName))
+				} else if stmt.stmtType == schemaCreateTable {
+					logger = log.With(zap.String("table", common.UniqueTable(job.dbName, stmt.tblName)))
+				} else if stmt.stmtType == schemaCreateView {
+					logger = log.With(zap.String("table", common.UniqueTable(job.dbName, stmt.tblName)))
 				}
-				err := InitSchema(ctx, rc.tidbGlue, dbMeta.Name, viewsSchema)
-
+				task := logger.Begin(zap.DebugLevel, fmt.Sprintf("execute SQL: %s", stmt.sql))
+				_, err := job.session.Execute(worker.ctx, stmt.sql)
 				task.End(zap.ErrorLevel, err)
 				if err != nil {
-					return errors.Annotatef(err, "restore view schema %s failed", dbMeta.Name)
+					switch stmt.stmtType {
+					case schemaCreateDatabase:
+						err = errors.Annotatef(err, "restore database schema %s failed", job.dbName)
+					case schemaCreateTable:
+						err = errors.Annotatef(err, "restore table schema %s failed", stmt.tblName)
+					case schemaCreateView:
+						err = errors.Annotatef(err, "restore view schema %s failed", stmt.tblName)
+					}
+					worker.wg.Done()
+					worker.throw(err)
 				}
 			}
+			worker.wg.Done()
+		}
+	}
+}
 
+func (worker *restoreSchemaWorker) wait() error {
+	defer func() {
+		// cancel whole jobs first, then close connectinos
+		worker.quit()
+		for _, session := range worker.sessions {
+			session.Close()
+		}
+	}()
+	select {
+	case err := <-worker.errCh:
+		return err
+	case <-worker.ctx.Done():
+		return nil
+	}
+}
+
+func (worker *restoreSchemaWorker) throw(err error) {
+	worker.errCh <- err
+}
+
+func (worker *restoreSchemaWorker) getSession(sessionID string) checkpoints.Session {
+	if _, opened := worker.sessions[sessionID]; !opened {
+		session, err := worker.glue.GetSession()
+		if err != nil {
+			worker.throw(err)
+		}
+		worker.sessions[sessionID] = session
+	}
+	return worker.sessions[sessionID]
+}
+
+func (rc *RestoreController) restoreSchema(ctx context.Context) error {
+	if !rc.cfg.Mydumper.NoSchema {
+		concurrency := 16
+		childCtx, cancel := context.WithCancel(ctx)
+		worker := restoreSchemaWorker{
+			ctx:      childCtx,
+			quit:     cancel,
+			wg:       sync.WaitGroup{},
```

Comment:
we could remove this implicit initiation of `WaitGroup`

Reply:
Sounds great!